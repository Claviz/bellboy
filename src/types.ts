import { CoreOptions, UriOptions, UrlOptions } from 'request';
import { RequestPromiseOptions } from 'request-promise';
import { ReadStream } from 'fs';
import { Readable } from 'stream';

export interface IDbConnection {
    user?: string;
    password?: string;
    server?: string;
    host?: string;
    database?: string;
    schema?: string;
}

export interface IReporter {
    report(job: IJob): void;
}

export interface IJobConfig {
    reporters?: IReporter[];
}

export interface IJob {
    on(eventName: string, cb: event): void;
    run(): Promise<void>;
    stop(message?: string): void;
    onAny(cb: event): void;
}

export interface IProcessorConfig {
    rowLimit?: number;
}

export interface IDynamicProcessorConfig extends IProcessorConfig {
    generator: () => AsyncIterableIterator<any>;
}

export interface IDirectoryProcessorConfig extends IProcessorConfig {
    path: string;
    filePattern?: string;
    files?: string[];
}

export interface IExcelProcessorConfig extends IDirectoryProcessorConfig {
    hasHeader?: boolean;
    ignoreEmpty?: boolean;
    sheets?: ((number | string)[] | sheetGetter);
}

export interface IProcessor {
    process(processStream: processStream, ...args: any): Promise<void>;
    rowLimit: number;
}

export interface IDestination {
    loadBatch: (data: any[]) => Promise<void>;
    batchSize: number;
    recordGenerator: ((row: any) => AsyncIterableIterator<{}>) | undefined;
    batchTransformer: ((rows: any[]) => Promise<any[]>) | undefined;
    // loadInMode: boolean;
    disableLoad: boolean;
}

export interface IDestinationConfig {
    disableLoad?: boolean;
    batchSize?: number;
    recordGenerator?: (row: any) => AsyncIterableIterator<{}>;
    batchTransformer?: (rows: any[]) => Promise<any[]>;
}

export interface IDatabaseDestinationConfig extends IDestinationConfig {
    table: string;
    connection: IDbConnection;
}

export interface IPostgresDestinationConfig extends IDatabaseDestinationConfig {
    upsertConstraints?: string[];
}

export interface IMssqlDestinationConfig extends IDatabaseDestinationConfig { }

export interface IHttpDestinationConfig extends IDestinationConfig {
    request: (UriOptions & RequestPromiseOptions) | (UrlOptions & RequestPromiseOptions);
}

export interface IStdoutDestinationConfig extends IDestinationConfig {
    asTable?: boolean;
}

export interface IHttpProcessorConfig extends IProcessorConfig {
    connection: CoreOptions & UrlOptions;
    nextRequest?: (header: any) => Promise<(CoreOptions & UrlOptions) | null | undefined>;
}

export interface IJsonHttpProcessorConfig extends IHttpProcessorConfig {
    dataFormat: 'json';
    jsonPath: string;
}

export interface IDelimitedHttpProcessorConfig extends IHttpProcessorConfig {
    dataFormat: 'delimited';
    delimiter: string;
}

export interface IDatabaseProcessorConfig extends IProcessorConfig {
    connection: any;
    query: string;
}

export interface IJsonProcessorConfig extends IDirectoryProcessorConfig {
    jsonPath: string;
}

export interface IDelimitedProcessorConfig extends IDirectoryProcessorConfig {
    delimiter: string;
}

export interface IMssqlProcessorConfig extends IDatabaseProcessorConfig {
}

export interface IPostgresProcessorConfig extends IDatabaseProcessorConfig {
}

export interface ITailProcessorConfig extends IDirectoryProcessorConfig {
    fromBeginning?: boolean;
}

export interface IMqttProcessorConfig extends IProcessorConfig {
    url: string;
    topics: string[];
}

export type DbTypes = 'postgres' | 'mssql';

export type event = (...args: any) => Promise<any>;

export type sheetGetter = (sheetNames: string[]) => Promise<string[]>;

export type emit = (event: string, ...args: any) => Promise<any>;

export type processStream = (readStream: Readable, ...args: any) => Promise<any>;
