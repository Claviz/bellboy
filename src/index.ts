export { Job } from './job';

export { Processor } from './processors/base/processor';
export { DatabaseProcessor } from './processors/base/database-processor';
export { DirectoryProcessor } from './processors/base/directory-processor';

export { ExcelProcessor } from './processors/excel-processor';
export { PostgresProcessor } from './processors/postgres-processor';
export { MssqlProcessor } from './processors/mssql-processor';
export { HttpProcessor } from './processors/http-processor';
export { DynamicProcessor } from './processors/dynamic-processor';
export { JsonProcessor } from './processors/json-processor';
export { MqttProcessor } from './processors/mqtt-processor';
export { TailProcessor } from './processors/tail-processor';
export { DelimitedProcessor } from './processors/delimited-processor';

export { Destination } from './destinations/base/destination';
export { DatabaseDestination } from './destinations/base/database-destination';

export { HttpDestination } from './destinations/http-destination';
export { MssqlDestination } from './destinations/mssql-destination';
export { PostgresDestination } from './destinations/postgres-destination';
export { StdoutDestination } from './destinations/stdout-destination';

export { getDb } from './utils';

export { Reporter } from './reporters/base/reporter'