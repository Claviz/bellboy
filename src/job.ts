import { Readable } from 'stream';

import { emit, event, IDestination, IJob, IJobConfig, IProcessor, IReporter } from './types';

export class Job implements IJob {

    protected closed: boolean = false;
    protected events: { [fn: string]: emit[] } = {};
    protected reporters: IReporter[];

    constructor(protected processor: IProcessor, protected destinations: IDestination[], config: IJobConfig = {}) {
        this.reporters = config.reporters || [];
        for (let i = 0; i < this.reporters.length; i++) {
            this.reporters[i].report(this);
        }
    }

    async run() {
        await this.emit('startProcessing', this.processor, this.destinations);
        try {
            const results: any[][] = [];
            for (let j = 0; j < this.destinations.length; j++) {
                results[j] = [];
            }
            let processedRowCount = 0;
            const processStreamWrapper = async (readStream: Readable, ...args: any[]) => {
                await this.emit('startProcessingStream', ...args);
                for await (const row of readStream) {
                    await this.emit('startProcessingRow', row);
                    for (let i = 0; i < this.destinations.length; i++) {
                        const recordGeneratorFn = this.destinations[i].recordGenerator;
                        if (!recordGeneratorFn) {
                            results[i].push(row);
                        } else {
                            const wrapper = async function* (recordGenerator: AsyncIterableIterator<{}>, emit: any) {
                                try {
                                    for await (const record of recordGenerator) {
                                        yield record;
                                    }
                                } catch (err) {
                                    await emit('rowGenerationError', i, row, err);
                                }
                            }
                            let recordGenerator = wrapper(recordGeneratorFn(row), (eventName: string, ...args: any) => this.emit(eventName, ...args));
                            for await (const record of recordGenerator) {
                                results[i].push(record);
                                await this.emit('rowGenerated', i, record);
                            }
                        }
                    }
                    await this.emit('endProcessingRow');
                    for (let i = 0; i < this.destinations.length; i++) {
                        while (this.destinations[i].batchSize && results[i].length >= this.destinations[i].batchSize) {
                            const destination = this.destinations[i];
                            const toSend = results[i].splice(0, destination.batchSize);
                            await this.loadBatch(i, destination, toSend);
                        }
                    }
                    processedRowCount++;
                    this.closed = this.processor.rowLimit === processedRowCount
                    if (this.closed) {
                        readStream.destroy();
                        break;
                    }
                }
                for (let i = 0; i < this.destinations.length; i++) {
                    if (results[i].length) {
                        await this.loadBatch(i, this.destinations[i], results[i]);
                    }
                    results[i] = [];
                }
                await this.emit('endProcessingStream', ...args);
            }
            await this.processor.process(processStreamWrapper);
            await this.emit('endProcessing');
        } catch (err) {
            await this.emit('processingError', err);
        }
    }

    on(eventName: 'endProcessing', fn: (() => Promise<any>)): any
    on(eventName: 'loadedBatch', fn: ((destinationIndex: number, data: any[]) => Promise<any>)): any
    on(eventName: 'processingError', fn: ((error: any) => Promise<any>)): any
    on(eventName: 'loadingBatchError', fn: ((destinationIndex: number, data: any[], error: any) => Promise<any>)): any
    on(eventName: 'loadingBatch', fn: ((destinationIndex: number, data: any[]) => Promise<any>)): any
    on(eventName: 'transformedBatch', fn: ((destinationIndex: number, data: any) => Promise<any>)): any
    on(eventName: 'transformingBatchError', fn: ((destinationIndex: number, data: any[], error: any) => Promise<any>)): any
    on(eventName: 'transformingBatch', fn: ((destinationIndex: number, data: any[]) => Promise<any>)): any
    on(eventName: 'endProcessingRow', fn: (() => Promise<any>)): any
    on(eventName: 'rowGenerationError', fn: ((destinationIndex: number, data: any, error: any) => Promise<any>)): any
    on(eventName: 'rowGenerated', fn: ((destinationIndex: number, row: any) => Promise<any>)): any
    on(eventName: 'startProcessingRow', fn: ((row: any) => Promise<any>)): any
    on(eventName: 'startProcessing', fn: (() => Promise<any>)): any
    on(eventName: 'endProcessingStream', fn: ((...args: any) => Promise<any>)): any
    on(eventName: 'startProcessingStream', fn: ((...args: any) => Promise<any>)): any
    on(eventName: string, fn: event) {
        const event = this.events[eventName];
        if (!event) {
            this.events[eventName] = [fn];
        } else {
            this.events[eventName].push(fn);
        }
    }

    protected async loadBatch(destinationIndex: number, destination: IDestination, data: any[]) {
        if (data.length) {
            if (destination.batchTransformer) {
                await this.emit('transformingBatch', destinationIndex, data);
                try {
                    data = await destination.batchTransformer(data);
                    await this.emit('transformedBatch', destinationIndex, data);
                } catch (err) {
                    await this.emit('transformingBatchError', destinationIndex, data, err);
                }
            }
            if (!destination.disableLoad) {
                await this.emit('loadingBatch', destinationIndex, data);
                try {
                    await destination.loadBatch(data);
                    await this.emit('loadedBatch', destinationIndex, data);
                } catch (err) {
                    await this.emit('loadingBatchError', destinationIndex, data, err);
                }
            }
        }
    }

    protected async emit(eventName: string, ...args: any) {
        const fn = this.events[eventName];
        if (fn) {
            for (let i = 0; i < fn.length; i++) {
                let result;
                try {
                    result = await fn[i].apply(this, args);
                } catch (err) { }
                if (result) {
                    this.closed = true;
                }
            }
        }
    }
}
