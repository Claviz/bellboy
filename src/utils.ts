import sql from 'mssql';
import { Readable } from 'stream';

import { DbTypes, IDbConnection } from './types';

const cachedDbConnections = new Map<string, { db: any, close: any }>();
export async function getDb(databaseConfig: IDbConnection, dbType: DbTypes) {
    const dbKey = JSON.stringify(databaseConfig);
    if (cachedDbConnections.has(dbKey)) {
        const dbConnection = cachedDbConnections.get(dbKey);
        if (dbConnection) {
            return dbConnection.db;
        }
    }
    if (dbType === 'mssql') {
        const pool = new sql.ConnectionPool({ ...databaseConfig } as sql.config);
        const db = await pool.connect();
        cachedDbConnections.set(dbKey, {
            db,
            close: pool.close.bind(pool),
        });
        return db;
    } else {
        const pgp = require('pg-promise')({ schema: databaseConfig.schema || 'public', });
        const db = pgp(databaseConfig);
        cachedDbConnections.set(dbKey, {
            db,
            close: pgp.end,
        });
        return db;
    }
};

export async function closeDbConnection(databaseConfig: IDbConnection) {
    const dbKey = JSON.stringify(databaseConfig);
    if (cachedDbConnections.has(dbKey)) {
        const dbConnection = cachedDbConnections.get(dbKey);
        if (dbConnection) {
            await dbConnection.close();
            cachedDbConnections.delete(dbKey);
        }
    }
}

export function getReadableJsonStream(jsonStream: Readable): Readable {
    const readableJsonStream = new Readable({
        objectMode: true,
        async read() {
            const data = await magic();
            if (!data) {
                return this.push(null);
            }
            this.push(data);
        },
    });
    const magic = async () => {
        return new Promise<any>(async (resolve, reject) => {
            function removeListeners() {
                jsonStream.removeListener('data', onRow);
                jsonStream.removeListener('done', onDone);
                jsonStream.removeListener('close', onDone);
                jsonStream.removeListener('error', onError);
            };
            async function onRow(row: any) {
                jsonStream.pause();
                removeListeners();
                resolve(row);
            };
            async function onDone() {
                removeListeners();
                resolve();
            };
            async function onError(err: any) {
                removeListeners();
                reject(err);
            };
            jsonStream.on('data', onRow);
            jsonStream.on('done', onDone);
            jsonStream.on('close', onDone);
            jsonStream.on('error', onError);
            jsonStream.resume();
        });
    }
    return readableJsonStream;
}