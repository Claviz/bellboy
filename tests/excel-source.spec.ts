import * as bellboy from '../src';
import { Destination } from '../src/types';
const xlsx = require('node-xlsx');
import * as fs from 'fs';

const filePath = 'test.xlsx';

beforeAll(async () => {

});

beforeEach(async () => {

});

afterAll(async () => {

});

afterEach(async () => {
    if (fs.existsSync(filePath)) {
        fs.unlinkSync(filePath);
    }
});

it('parses xlsx without header', async () => {
    let data: any[] = [];
    const buffer = xlsx.build([{
        name: 'sheet1',
        data: [['hello', 'world']],
    }]);
    fs.writeFileSync(filePath, buffer);

    const processor = new bellboy.ExcelProcessor({
        hasHeader: false,
        path: './',
        files: [filePath],
        destinations: [
            {
                type: 'custom',
                batchSize: 1,
                load: async (rows) => {
                    data = [...data, ...rows];
                }
            } as Destination
        ],

    });
    await processor.process();
    expect(data).toEqual([{
        A: 'hello',
        B: 'world',
    }]);
});

it('parses xlsx with header', async () => {
    let data: any[] = [];
    const buffer = xlsx.build([{
        name: 'sheet1',
        data: [['column1', 'column2'], ['hello', 'world']],
    }]);
    fs.writeFileSync(filePath, buffer);

    const processor = new bellboy.ExcelProcessor({
        hasHeader: true,
        path: './',
        files: [filePath],
        destinations: [
            {
                type: 'custom',
                batchSize: 1,
                load: async (rows) => {
                    data = [...data, ...rows];
                }
            } as Destination
        ],

    });
    await processor.process();
    expect(data).toEqual([{
        column1: 'hello',
        column2: 'world',
    }]);
});

it('parses all xlsx files by pattern', async () => {
    let data: any[] = [];
    const buffer = xlsx.build([{
        name: 'sheet1',
        data: [['test']],
    }]);
    fs.writeFileSync(filePath, buffer);

    const processor = new bellboy.ExcelProcessor({
        hasHeader: false,
        path: './',
        filePattern: `.*\.(xlsx)`,
        destinations: [
            {
                type: 'custom',
                batchSize: 1,
                load: async (rows) => {
                    data = [...data, ...rows];
                }
            } as Destination
        ],

    });
    await processor.process();
    expect(data).toEqual([{
        A: 'test',
    }]);
});

it('parses specific sheet by name', async () => {
    let data: any[] = [];
    const buffer = xlsx.build([{
        name: 'sheet1',
        data: [['test1']],
    }, {
        name: 'sheet2',
        data: [['test2']],
    }, {
        name: 'sheet3',
        data: [['test3']],
    }]);
    fs.writeFileSync(filePath, buffer);

    const processor = new bellboy.ExcelProcessor({
        hasHeader: false,
        path: './',
        files: [filePath],
        sheetName: 'sheet2',
        destinations: [
            {
                type: 'custom',
                batchSize: 1,
                load: async (rows) => {
                    data = [...data, ...rows];
                }
            } as Destination
        ],

    });
    await processor.process();
    expect(data).toEqual([{
        A: 'test2',
    }]);
});

it('parses specific sheet by index', async () => {
    let data: any[] = [];
    const buffer = xlsx.build([{
        name: 'sheet1',
        data: [['test1']],
    }, {
        name: 'sheet2',
        data: [['test2']],
    }, {
        name: 'sheet3',
        data: [['test3']],
    }]);
    fs.writeFileSync(filePath, buffer);

    const processor = new bellboy.ExcelProcessor({
        hasHeader: false,
        path: './',
        files: [filePath],
        sheetIndex: 2,
        destinations: [
            {
                type: 'custom',
                batchSize: 1,
                load: async (rows) => {
                    data = [...data, ...rows];
                }
            } as Destination
        ],

    });
    await processor.process();
    expect(data).toEqual([{
        A: 'test3',
    }]);
});