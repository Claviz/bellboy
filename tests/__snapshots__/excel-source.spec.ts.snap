// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses all xlsx files by pattern 1`] = `
Array [
  Object {
    "formatted": Object {
      "arr": Array [
        "test",
      ],
      "obj": Object {
        "A": "test",
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        "test",
      ],
      "obj": Object {
        "A": "test",
      },
    },
  },
]
`;

exports[`parses multiple sheets 1`] = `
Array [
  Object {
    "formatted": Object {
      "arr": Array [
        "test3",
      ],
      "obj": Object {
        "A": "test3",
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        "test3",
      ],
      "obj": Object {
        "A": "test3",
      },
    },
  },
  Object {
    "formatted": Object {
      "arr": Array [
        "test2",
      ],
      "obj": Object {
        "A": "test2",
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        "test2",
      ],
      "obj": Object {
        "A": "test2",
      },
    },
  },
  Object {
    "formatted": Object {
      "arr": Array [
        "test1",
      ],
      "obj": Object {
        "A": "test1",
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        "test1",
      ],
      "obj": Object {
        "A": "test1",
      },
    },
  },
]
`;

exports[`parses specific sheet by function 1`] = `
Array [
  Object {
    "formatted": Object {
      "arr": Array [
        "test3",
      ],
      "obj": Object {
        "A": "test3",
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        "test3",
      ],
      "obj": Object {
        "A": "test3",
      },
    },
  },
]
`;

exports[`parses specific sheet by index 1`] = `
Array [
  Object {
    "formatted": Object {
      "arr": Array [
        "test3",
      ],
      "obj": Object {
        "A": "test3",
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        "test3",
      ],
      "obj": Object {
        "A": "test3",
      },
    },
  },
]
`;

exports[`parses specific sheet by name 1`] = `
Array [
  Object {
    "formatted": Object {
      "arr": Array [
        "test2",
      ],
      "obj": Object {
        "A": "test2",
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        "test2",
      ],
      "obj": Object {
        "A": "test2",
      },
    },
  },
]
`;

exports[`parses xlsx with header 1`] = `
Array [
  Object {
    "formatted": Object {
      "arr": Array [
        "hello",
        "world",
      ],
      "obj": Object {
        "column1": "hello",
        "column2": "world",
      },
    },
    "header": Array [
      "column1",
      "column2",
    ],
    "raw": Object {
      "arr": Array [
        "hello",
        "world",
      ],
      "obj": Object {
        "column1": "hello",
        "column2": "world",
      },
    },
  },
]
`;

exports[`parses xlsx without header 1`] = `
Array [
  Object {
    "formatted": Object {
      "arr": Array [
        "hello",
        "world",
      ],
      "obj": Object {
        "A": "hello",
        "B": "world",
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        "hello",
        "world",
      ],
      "obj": Object {
        "A": "hello",
        "B": "world",
      },
    },
  },
]
`;

exports[`respects rowLimit 1`] = `
Array [
  Object {
    "formatted": Object {
      "arr": Array [
        1,
      ],
      "obj": Object {
        "A": 1,
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        1,
      ],
      "obj": Object {
        "A": 1,
      },
    },
  },
  Object {
    "formatted": Object {
      "arr": Array [
        2,
      ],
      "obj": Object {
        "A": 2,
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        2,
      ],
      "obj": Object {
        "A": 2,
      },
    },
  },
  Object {
    "formatted": Object {
      "arr": Array [
        3,
      ],
      "obj": Object {
        "A": 3,
      },
    },
    "header": Array [],
    "raw": Object {
      "arr": Array [
        3,
      ],
      "obj": Object {
        "A": 3,
      },
    },
  },
]
`;
